src/jobservice/job/impl/gc/garbage_collection.go

```diff
@@ -423,17 +423,33 @@ func (gc *GarbageCollector) removeUntaggedBlobs(ctx job.Context) {
 			continue
 		}
 		p := result.Data
-		all, err := gc.blobMgr.List(ctx.SystemContext(), blob.ListParams{
-			ProjectID:  p.ProjectID,
-			UpdateTime: time.Now().Add(-time.Duration(gc.timeWindowHours) * time.Hour),
-		})
-		if err != nil {
-			gc.logger.Errorf("failed to get blobs of project, %v", err)
-			continue
-		}
-		if err := gc.blobMgr.CleanupAssociationsForProject(ctx.SystemContext(), p.ProjectID, all); err != nil {
-			gc.logger.Errorf("failed to clean untagged blobs of project, %v", err)
-			continue
+
+		pn := int64(1)
+		for {
+			rg := q.Range{
+				Max: time.Now().Add(-time.Duration(gc.timeWindowHours) * time.Hour).Format(time.RFC3339),
+			}
+			q := &q.Query{
+				Keywords: map[string]interface{}{
+					"update_time": &rg,
+					"projectID":   p.ProjectID,
+				},
+				PageNumber: pn,
+				PageSize:   10000,
+			}
+			all, err := gc.blobMgr.List(ctx.SystemContext(), q)
+			if err != nil {
+				gc.logger.Errorf("failed to get blobs of project, %v", err)
+				break
+			}
+			if err := gc.blobMgr.CleanupAssociationsForProject(ctx.SystemContext(), p.ProjectID, all); err != nil {
+				gc.logger.Errorf("failed to clean untagged blobs of project, %v", err)
+				break
+			}
+			if len(all) < 10000 {
+				break
+			}
+			pn++
```

Comment:
There are blobs 1, 2, 3, 4, 5, 6, 7, 9, 10 in the project, we list the blobs with a page size of 2. And 1, 2, 3, 4 will be cleaned in GC.

Well, we will get 1, 2 when pg is 1, then we delete the blobs 1 and 2, pg increase to  2. The next time, we get the 5 and 6, blob 3 and 4 are skipped, this is wrong.

We need to ensure to get blob 3 and 4 even the blob 1 and 2 are deleted before.

Suggest to process like the following

```go

pageSize = 1000
lastBlobID := 0

q := &q.Query{
	Keywords: map[string]interface{}{
		"update_time": &rg,
		"projectID":   p.ProjectID,
		"id__gt" lastBlobID,
	},
	PageSize: int64(pageSize),
	PageNumber: 1,
	Sorting: "id"
}

for {
	blobs, err := gc.blobMgr.List(ctx.SystemContext(), q)
	if err != nil {
		gc.logger.Errorf("failed to get blobs of project, %v", err)
		break
	}
	if err := gc.blobMgr.CleanupAssociationsForProject(ctx.SystemContext(), p.ProjectID, blobs); err != nil {
		gc.logger.Errorf("failed to clean untagged blobs of project, %v", err)
		break
	}

	if len(blobs) < pageSize {
		break
	}

	lastBlobID = blobs[len(blobs) -1].ID
}

```

Reply:
here maybe a misunderstanding, this step will not remove data from table blob but project_blob. So, the `all` is static within the for loop.

Reply:
discussed offline, mark it as resolved.

Reply:
good catch, buddy